x/gamm/pool-models/internal/cfmm_common/lp.go

```diff
@@ -0,0 +1,163 @@
+		shareRatio := coin.Amount.ToDec().QuoInt(poolLiquidity.AmountOfNoDenomValidation(coin.Denom))
+		if shareRatio.LT(minShareRatio) {
+			minShareRatio = shareRatio
+		}
+		if shareRatio.GT(maxShareRatio) {
+			maxShareRatio = shareRatio
+		}
+		coinShareRatios[i] = shareRatio
+	}
+
+	if minShareRatio.Equal(sdk.MaxSortableDec) {
+		return numShares, remCoins, errors.New("unexpected error in MaximalExactRatioJoin")
+	}
+
+	remCoins = sdk.Coins{}
+	numShares = minShareRatio.MulInt(totalShares).TruncateInt()
+
+	// if we have multiple share values, calculate remainingCoins
+	if !minShareRatio.Equal(maxShareRatio) {
+		// we have to calculate remCoins
+		for i, coin := range tokensIn {
+			// if coinShareRatios[i] == minShareRatio, no remainder
+			if coinShareRatios[i].Equal(minShareRatio) {
+				continue
+			}
+
+			usedAmount := minShareRatio.MulInt(coin.Amount).Ceil().TruncateInt()
+			newAmt := coin.Amount.Sub(usedAmount)
+			// if newAmt is non-zero, add to RemCoins. (It could be zero due to rounding)
+			if !newAmt.IsZero() {
+				remCoins = remCoins.Add(sdk.Coin{Denom: coin.Denom, Amount: newAmt})
+			}
+		}
+	}
+
+	return numShares, remCoins, nil
+}
+
+// We binary search a number of LP shares, s.t. if we exited the pool with the updated liquidity,
+// and swapped all the tokens back to the input denom, we'd get the same amount. (under 0 swap fee)
+// Thanks to CFMM path-independence, we can estimate slippage with these swaps to be sure to get the right numbers here.
+// (by path-independence, swap all of B -> A, and then swap all of C -> A will yield same amount of A, regardless
+// of order and interleaving)
+//
+// This implementation all of pool.GetTotalPoolLiquidity, pool.ExitPool, and pool.SwapExactAmountIn to not
+// require any updates to state, and instead only do updates based upon the pool struct.
+func BinarySearchSingleAssetJoin(
+	pool types.PoolI,
+	tokenIn sdk.Coin,
+	poolWithAddedLiquidityAndShares func(newLiquidity sdk.Coin, newShares sdk.Int) types.PoolI,
+) (numLPShares sdk.Int, err error) {
+	// use dummy context
+	ctx := sdk.Context{}
+	// Need to get something that makes the result correct within 1 LP share
+	// If we fail to reach it within maxIterations, we return an error
+	correctnessThreshold := sdk.NewInt(2)
+	maxIterations := 512
+	// upperbound of number of LP shares = existingShares * tokenIn.Amount / pool.totalLiquidity.AmountOf(tokenIn.Denom)
+	existingTokenLiquidity := pool.GetTotalPoolLiquidity(ctx).AmountOf(tokenIn.Denom)
+	existingLPShares := pool.GetTotalShares()
+	LPShareUpperBound := existingLPShares.Mul(tokenIn.Amount).ToDec().QuoInt(existingTokenLiquidity).Ceil().TruncateInt()
+	LPShareLowerBound := sdk.ZeroInt()
+
+	// Creates a pool with tokenIn liquidity added, where it created `sharesIn` number of shares.
+	// Returns how many tokens you'd get, if you then exited all of `sharesIn` for tokenIn.Denom
+	estimateCoinOutGivenShares := func(sharesIn sdk.Int) (tokenOut sdk.Int, err error) {
```

Comment:
What do you think about making this an actual function so that it can be unit tested?

Reply:
+1 on this comment!